File: aoc_5.R
read_coord <- function(path) {
  x <- gsub(",| -> ", " ", readLines(path))
  read.table(text = x, col.names = c("x1", "y1", "x2", "y2"))
}

count_overlaps <- function(x1, y1, x2, y2) {
  x <- Map(seq, x1, x2)
  y <- Map(seq, y1, y2)
  coord <- do.call(rbind, Map(cbind, x, y))
  nrow(unique(coord[duplicated(coord), ]))
}

solve <- function(path, part1 = TRUE) {
  d <- read_coord(path)
  if (part1) {
    straight <- with(d, x1 == x2 | y1 == y2)
    d <- d[straight, ]
  }
  with(d, count_overlaps(x1, y1, x2, y2))
}

c(solve("data/aoc_5"),
  solve("data/aoc_5", part1 = FALSE))

File: aoc_5.jl
function parse_coordinates(x)
    x1, y1, x2, y2 = parse.(Int, split(x, r",| -> "))
    xstep = x1 > x2 ? -1 : 1
    ystep = y1 > y2 ? -1 : 1
    CartesianIndex.(x1:xstep:x2, y1:ystep:y2)
end

# function solve(path)
#     d = [parse_coordinates(lines) for lines in readlines(path)]
#     d = [(d...)...]
#     sum([count(==(i), d) for i in unique(d)] .> 1) # crazy slow
# end

# println(solve("data/aoc_5"))

function solve_faster(path)
    coord = parse_coordinates.(readlines(path))
    x = zeros(Int, 1000, 1000)
    for i = coord
        x[i] .+= 1
    end
    sum(x .> 1)
end

println(solve_faster("data/aoc_5"))

